File: src/ios/ACPCore_Cordova.m
/*
 Copyright 2020 Adobe. All rights reserved.
 This file is licensed to you under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License. You may obtain a copy
 of the License at http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software distributed under
 the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 OF ANY KIND, either express or implied. See the License for the specific language
 governing permissions and limitations under the License.
 */

 

#import <ACPCore/ACPCore.h>
#import <ACPCore/ACPExtensionEvent.h>
#import <ACPCore/ACPIdentity.h>
#import <ACPCore/ACPLifecycle.h>
#import <ACPCore/ACPSignal.h>
#import <ACPMobileServices/ACPMobileServices.h>
#import <ACPTarget/ACPTarget.h>
#import <ACPAnalytics/ACPAnalytics.h>
#import <ACPUserProfile/ACPUserProfile.h>
#import <Cordova/CDV.h>
#import <Foundation/Foundation.h>



@import UserNotifications;

//#import <FirebaseCore/FIRApp.h>
//#import <FirebaseMessaging/FirebaseMessaging.h>
//@import Firebase;



@interface ACPCore_Cordova : CDVPlugin
- (void) dispatchEvent:(CDVInvokedUrlCommand*)command;
- (void) dispatchEventWithResponseCallback:(CDVInvokedUrlCommand*)command;
- (void) dispatchResponseEvent:(CDVInvokedUrlCommand*)command;
- (void) downloadRules:(CDVInvokedUrlCommand*)command;
- (void) extensionVersion:(CDVInvokedUrlCommand*)command;
- (void) getPrivacyStatus:(CDVInvokedUrlCommand*)command;
- (void) getSdkIdentities:(CDVInvokedUrlCommand*)command;
- (void) setAdvertisingIdentifier:(CDVInvokedUrlCommand*)command;
- (void) setLogLevel:(CDVInvokedUrlCommand*)command;
- (void) setPrivacyStatus:(CDVInvokedUrlCommand*)command;
- (void) trackAction:(CDVInvokedUrlCommand*)command;
- (void) trackState:(CDVInvokedUrlCommand*)command;
- (void) updateConfiguration:(CDVInvokedUrlCommand*)command;
- (void) getAppId:(CDVInvokedUrlCommand*)command;
- (void) setPushIdentifier:(CDVInvokedUrlCommand*)command;
- (void)loadAdobe:(CDVInvokedUrlCommand*)command;

- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken;
- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error;
- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult result))handler;
- (void)applicationDidBecomeActive:(UIApplication *)application;

@end

@implementation ACPCore_Cordova

 NSString *appId;
 NSString *initTime;

- (void) dispatchEvent:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSDictionary *eventInput = [self getCommandArg:command.arguments[0]];
        if (![eventInput isKindOfClass:[NSDictionary class]]) {
            [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Unable to dispatch event. Input was malformed"] callbackId:command.callbackId];
        }

        ACPExtensionEvent *event = [self getExtensionEventFromJavascriptObject:eventInput];
        NSError *error = nil;
        [ACPCore dispatchEvent:event error:&error];

        if (error) {
            [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Error dispatching event: %@", error.localizedDescription ?: @"unknown error"]] callbackId:command.callbackId];
        }

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) dispatchEventWithResponseCallback:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSDictionary *eventInput = [self getCommandArg:command.arguments[0]];
        if (![eventInput isKindOfClass:[NSDictionary class]]) {
            [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Unable to dispatch event. Input was malformed"] callbackId:command.callbackId];
        }

        ACPExtensionEvent *event = [self getExtensionEventFromJavascriptObject:eventInput];
        NSError *error = nil;
        [ACPCore dispatchEventWithResponseCallback:event
                                  responseCallback:^(ACPExtensionEvent * _Nonnull responseEvent) {
            if (error) {
                [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Error dispatching event: %@", error.localizedDescription ?: @"unknown error"]] callbackId:command.callbackId];
            }

            NSDictionary *response = [self getJavascriptDictionaryFromEvent:responseEvent];
            CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:response];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }
                                             error:&error];
    }];
}

- (void) dispatchResponseEvent:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSDictionary *inputResponseEvent = [self getCommandArg:command.arguments[0]];
        NSDictionary *inputRequestEvent = [self getCommandArg:command.arguments[1]];

        if (![inputRequestEvent isKindOfClass:[NSDictionary class]] || ![inputResponseEvent isKindOfClass:[NSDictionary class]]) {
            [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Unable to dispatch event. Input was malformed"] callbackId:command.callbackId];
        }

        ACPExtensionEvent *responseEvent = [self getExtensionEventFromJavascriptObject:inputResponseEvent];
        ACPExtensionEvent *requestEvent = [self getExtensionEventFromJavascriptObject:inputRequestEvent];
        NSError *error = nil;
        [ACPCore dispatchResponseEvent:responseEvent
                          requestEvent:requestEvent
                                 error:&error];

        if (error) {
            [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Error dispatching response event: %@", error.localizedDescription ?: @"unknown error"]] callbackId:command.callbackId];
        }

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) downloadRules:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        [ACPCore downloadRules];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) extensionVersion:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSString *version = [[initTime stringByAppendingString:@": "] stringByAppendingString:[ACPCore extensionVersion]];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:version];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) getPrivacyStatus:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        [ACPCore getPrivacyStatus:^(ACPMobilePrivacyStatus status) {
            CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsNSInteger:status];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    }];
}

- (void) getSdkIdentities:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        [ACPCore getSdkIdentities:^(NSString * _Nullable content) {
            CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:content];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    }];
}

- (void) setAdvertisingIdentifier:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSString *newIdentifier = [self getCommandArg:command.arguments[0]];

        [ACPCore setAdvertisingIdentifier:newIdentifier];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) setLogLevel:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        ACPMobileLogLevel logLevel = (ACPMobileLogLevel)[[self getCommandArg:command.arguments[0]] intValue];

        [ACPCore setLogLevel:logLevel];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) setPrivacyStatus:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        ACPMobilePrivacyStatus privacyStatus;
	    ACPMobilePrivacyStatus privacyStatus = (ACPMobilePrivacyStatus)[[self getCommandArg:command.arguments[0]] intValue];
        /*switch ([self getCommandArg:command.arguments[0]] intValue) {
            case 0:
                privacyStatus = ACPMobilePrivacyStatusOptIn;
		break;
            case 1:
                privacyStatus = ACPMobilePrivacyStatusOptOut;
		break;
	    case 2:
            default:
                privacyStatus = ACPMobilePrivacyStatusUnknown;
		break;
        }*/

        [ACPCore setPrivacyStatus:privacyStatus];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) trackAction:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        id firstArg = [self getCommandArg:command.arguments[0]];
        id secondArg = [self getCommandArg:command.arguments[1]];

        // allows the ACPCore.trackAction(cData) call
        if([firstArg isKindOfClass:[NSDictionary class]]) {
            [ACPCore trackAction:nil data:firstArg];
        }
        else {
            [ACPCore trackAction:firstArg data:secondArg];
        }

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) trackState:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        id firstArg = [self getCommandArg:command.arguments[0]];
        id secondArg = [self getCommandArg:command.arguments[1]];

        // allows the ACPCore.trackState(cData) call
        if([firstArg isKindOfClass:[NSDictionary class]]) {
            [ACPCore trackState:nil data:firstArg];
        }
        else {
            [ACPCore trackState:firstArg data:secondArg];
        }

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) updateConfiguration:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSDictionary *config = [self getCommandArg:command.arguments[0]];

        [ACPCore updateConfiguration:config];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) getAppId:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
         CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:appId];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) setPushIdentifier:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
       NSString *token = [self getCommandArg:command.arguments[0]];

        [ACPCore setPushIdentifier:token];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

// ===========================================================================
// helper functions
// ===========================================================================
- (ACPExtensionEvent *) getExtensionEventFromJavascriptObject:(NSDictionary *)event {
    NSError *error = nil;
    ACPExtensionEvent *newEvent = [ACPExtensionEvent extensionEventWithName:event[@"name"]
                                                                       type:event[@"type"]
                                                                     source:event[@"source"]
                                                                       data:event[@"data"]
                                                                      error:&error];
    if (error || !newEvent) {
        [ACPCore log:ACPMobileLogLevelWarning tag:@"ACPCore" message:[NSString stringWithFormat:@"Error creating ACPExtensionEvent: %@", error.localizedDescription ?: @"unknown"]];
    }

    return newEvent;
}

- (NSDictionary *) getJavascriptDictionaryFromEvent:(ACPExtensionEvent *)event {
    return @{
        @"name" : event.eventName,
        @"type" : event.eventType,
        @"source" : event.eventSource,
        @"data" : event.eventData
    };
}

- (id) getCommandArg:(id) argument {
    return argument == (id)[NSNull null] ? nil : argument;
}

// ===============================================================
// Plugin lifecycle events
// ===============================================================
- (void)pluginInitialize
{
    appId = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"AppId"];

    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];  
      center.delegate = self;  
      [center requestAuthorizationWithOptions:(UNAuthorizationOptionSound | UNAuthorizationOptionAlert | UNAuthorizationOptionBadge) completionHandler:^(BOOL granted, NSError * _Nullable error) {
        if( !error ) {
            // required to get the app to do anything at all about push notifications  
            dispatch_async(dispatch_get_main_queue(), ^{
                [[UIApplication sharedApplication] registerForRemoteNotifications];
            });
            NSLog( @"Push registration success." );  
        } else {
            NSLog( @"Push registration FAILED" );  
            NSLog( @"ERROR: %@ - %@", error.localizedFailureReason, error.localizedDescription );  
            NSLog( @"SUGGESTIONS: %@ - %@", error.localizedRecoveryOptions, error.localizedRecoverySuggestion );  
        }
        }];
      
    [ACPCore setLogLevel:ACPMobileLogLevelDebug];
    
    [ACPCore setWrapperType:ACPMobileWrapperTypeCordova];
    [ACPCore configureWithAppId:appId];
    
    [ACPAnalytics registerExtension];
    [ACPMobileServices registerExtension];
    [ACPTarget registerExtension];
    [ACPIdentity registerExtension];
    [ACPLifecycle registerExtension];
    [ACPSignal registerExtension];
    [ACPUserProfile registerExtension];


    [ACPCore start:^{
        [ACPCore lifecycleStart:nil];
        [ACPCore collectPii:@{@"cusFiscalNumber": @"111111111"}];
    }];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
   // [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
 
}

- (void)loadAdobe:(CDVInvokedUrlCommand*)command
{

    appId = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"AppId"];
    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];  
      center.delegate = self;  
      [center requestAuthorizationWithOptions:(UNAuthorizationOptionSound | UNAuthorizationOptionAlert | UNAuthorizationOptionBadge) completionHandler:^(BOOL granted, NSError * _Nullable error) {
        if(!error) {
            // required to get the app to do anything at all about push notifications  
            dispatch_async(dispatch_get_main_queue(), ^{
                [[UIApplication sharedApplication] registerForRemoteNotifications];
            });
            NSLog( @"Push registration success." );  
        } else {
            NSLog( @"Push registration FAILED" );  
            NSLog( @"ERROR: %@ - %@", error.localizedFailureReason, error.localizedDescription );  
            NSLog( @"SUGGESTIONS: %@ - %@", error.localizedRecoveryOptions, error.localizedRecoverySuggestion );  
        }
        }];
      
    [ACPCore setLogLevel:ACPMobileLogLevelDebug];
    
    [ACPCore setWrapperType:ACPMobileWrapperTypeCordova];

    [ACPCore configureWithAppId:appId];
    
    [ACPAnalytics registerExtension];
    [ACPMobileServices registerExtension];
    [ACPTarget registerExtension];
    [ACPIdentity registerExtension];
    [ACPLifecycle registerExtension];
    [ACPSignal registerExtension];
    [ACPUserProfile registerExtension];


    [ACPCore start:^{
        [ACPCore lifecycleStart:nil];
        [ACPCore collectPii:@{@"cusFiscalNumber": @"111111111"}];
    }];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
   // [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
    NSLog(@"setPushIdentifier: %@", deviceToken);
    [ACPCore setPushIdentifier:deviceToken];
        

}

- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error {
    NSLog(@"didFailToRegisterForRemoteNotificationsWithError");
}

- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult result))handler
{
    NSLog(@"didReceiveNotification");
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
    application.applicationIconBadgeNumber = 0;
}

@end

File: src/ios/ACPCore_Cordova.swift
import AEPAnalytics
import AEPCampaign
import AEPCore
import AEPMobileServices
import AEPPlaces
import AEPServices
import AEPTarget
import AEPUserProfile

@objc(ACPCore_Cordova) class ACPCore_Cordova: CDVPlugin {

  var appId: String!
  var initTime: String!

  @objc(dispatchEvent:)
  func dispatchEvent(command: CDVInvokedUrlCommand) {
      self.commandDelegate.run(inBackground: {
          guard let eventInput = command.arguments[0] as? [String: Any],
                let name = eventInput["name"] as? String,
                let type = eventInput["type"] as? String,
                let source = eventInput["source"] as? String else {
              let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid event data")
              self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
              return
          }
          let event = Event(name: name, type: type, source: source, data: eventInput["data"] as? [String: String])
          MobileCore.dispatch(event: event)
          let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
          self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
      })
  }

  @objc(dispatchEventWithResponseCallback:)
  func dispatchEventWithResponseCallback(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      guard let eventInput = command.arguments[0] as? NSDictionary else {
        self.commandDelegate.send(
          CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: "Unable to dispatch event. Input was malformed"),
          callbackId: command.callbackId)
        return
      }

      let event: AEPCore.Event! = self.getExtensionEventFromJavascriptObject(event: eventInput)

      MobileCore.dispatch(
        event: event,
        responseCallback: { (response: AEPCore.Event!) in
          let responseEvent: NSDictionary! = self.getJavascriptDictionaryFromEvent(
            event: response)
          let pluginResult: CDVPluginResult! = CDVPluginResult(
            status: CDVCommandStatus_OK, messageAs: responseEvent as? [AnyHashable: Any])
          self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        })
    })
  }

  @objc(dispatchResponseEvent:)
  func dispatchResponseEvent(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      guard let inputResponseEvent = command.arguments[0] as? NSDictionary else {
        self.commandDelegate.send(
          CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: "Unable to dispatch event. InputResponse was malformed"),
          callbackId: command.callbackId)
        return
      }

      guard let inputRequestEvent = command.arguments[1] as? NSDictionary else {
        self.commandDelegate.send(
          CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: "Unable to dispatch event. InputResquest was malformed"),
          callbackId: command.callbackId)
        return
      }

      self.commandDelegate.send(
        CDVPluginResult(
          status: CDVCommandStatus_ERROR,
          messageAs: "Deprecated"),
        callbackId: command.callbackId)

    })
  }

  @objc(downloadRules:)
  func downloadRules(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_ERROR)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(extensionVersion:)
  func extensionVersion(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      let version: String! = self.initTime.appending(": ").appending(MobileCore.extensionVersion)
      let pluginResult: CDVPluginResult! = CDVPluginResult(
        status: CDVCommandStatus_OK, messageAs: version)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(getPrivacyStatus:)
  func getPrivacyStatus(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      MobileCore.getPrivacyStatus { PrivacyStatus in
        let pluginResult: CDVPluginResult! = CDVPluginResult(
          status: CDVCommandStatus_OK, messageAs: PrivacyStatus.rawValue)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
      }
    })
  }

  @objc(getSdkIdentities:)
  func getSdkIdentities(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      MobileCore.getSdkIdentities { content, error in
        if error == nil {
          let pluginResult: CDVPluginResult! = CDVPluginResult(
            status: CDVCommandStatus_OK, messageAs: content)
          self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
      }
    })
  }

  @objc(setAdvertisingIdentifier:)
  func setAdvertisingIdentifier(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      let newIdentifier: String! = command.arguments[0] as? String
      MobileCore.setAdvertisingIdentifier(newIdentifier)

      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(setLogLevel:)
  func setLogLevel(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      let logLevel: AEPServices.LogLevel =
        command.arguments[0] as? AEPServices.LogLevel ?? AEPServices.LogLevel.warning

      MobileCore.setLogLevel(logLevel)

      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(setPrivacyStatus:)
  func setPrivacyStatus(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      let privacyStatus: PrivacyStatus =
        command.arguments[0] as? PrivacyStatus ?? PrivacyStatus.unknown

      MobileCore.setPrivacyStatus(privacyStatus)
      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(trackAction:)
  func trackAction(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      let firstArg: AnyObject! = command.arguments[0] as AnyObject
      let secondArg: AnyObject! = command.arguments[1] as AnyObject

      // allows the AEPCore.trackAction(cData) call
      if firstArg is NSDictionary {
        MobileCore.track(action: nil, data: firstArg as? [String: String])
      } else {
        MobileCore.track(action: firstArg as? String, data: secondArg as? [String: String])
      }

      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(trackState:)
  func trackState(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      let firstArg: AnyObject! = command.arguments[0] as AnyObject
      let secondArg: AnyObject! = command.arguments[1] as AnyObject

      // allows the AEPCore.trackAction(cData) call
      if firstArg is NSDictionary {
        MobileCore.track(state: nil, data: firstArg as? [String: String])
      } else {
        MobileCore.track(state: firstArg as? String, data: secondArg as? [String: String])
      }

      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(updateConfiguration:)
  func updateConfiguration(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      guard let config = command.arguments[0] as? [String: Any] else {
        self.commandDelegate.send(
          CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: "Unable to dispatch event. InputResponse was malformed"),
          callbackId: command.callbackId)
        return
      }

      MobileCore.updateConfigurationWith(configDict: config)

      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(getAppId:)
  func getAppId(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      let pluginResult: CDVPluginResult! = CDVPluginResult(
        status: CDVCommandStatus_OK, messageAs: self.appId)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }


    @objc(openDeepLink:)
    func openDeepLink(command: CDVInvokedUrlCommand!) {
      self.commandDelegate.run(inBackground: {
          ACPAppDelegatePush.openScreenByDeepLink(command.arguments[0] as! String)
      })
    }

  // ===========================================================================
  // helper functions
  // ===========================================================================

  func getExtensionEventFromJavascriptObject(event: NSDictionary!) -> AEPCore.Event! {

    let newEvent = AEPCore.Event.init(
      name: event.value(forKey: "name") as! String,
      type: event.value(forKey: "type") as! String,
      source: event.value(forKey: "source") as! String,
      data: event.value(forKey: "data") as? [String: String])

    return newEvent
  }

  func getJavascriptDictionaryFromEvent(event: AEPCore.Event!) -> NSDictionary! {
    return [
      "name": event.name,
      "type": event.type,
      "source": event.source,
      "data": event.data!,
    ]
  }

  // ===============================================================
  // Plugin lifecycle events
  // ===============================================================
  override func pluginInitialize() {

    let date: NSDate! = NSDate()
    let dateFormatter: DateFormatter! = DateFormatter()
    dateFormatter.dateFormat = "dd/MM/yyyy HH:mm:ss"
    initTime = dateFormatter.string(from: date as Date)
    self.appId = Bundle.main.object(forInfoDictionaryKey: "AppId") as? String
    ACPAppDelegatePush.registerExtensions()
  }
}
